releng/release-tool/release-tool.go

```diff
@@ -0,0 +1,1186 @@
+			if *release.TagName == r.promoteRC {
+				found = true
+				invalid, _, err = r.isRCInvalid(release, expectedBranch)
+				if err != nil {
+					return err
+				}
+				r.promoteRCTime = release.CreatedAt.Time
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("Unable to find promote-rc tag [%s]", r.promoteRC)
+		} else if invalid {
+			return fmt.Errorf("RC [%s] is ineligible to be promoted due to blockers.", r.promoteRC)
+		}
+	} else {
+		// if this is not a promotion, ensure the release is either an RC or a patch release.
+		re := regexp.MustCompile(`^v\d*\.\d*.\d*-rc.\d*$`)
+		match := re.FindString(r.tag)
+		if match == "" && tagSemver.Patch() == 0 && !r.force {
+			return fmt.Errorf("The tag [%s] must be promoted from a release candidate since it is the first release of a patch series.", r.tag)
+		}
+
+	}
+
+	var vs []*semver.Version
+
+	for _, release := range releases {
+		if *release.Draft ||
+			*release.Prerelease ||
+			len(release.Assets) == 0 {
+
+			continue
+		}
+		v, err := semver.NewVersion(*release.TagName)
+		if err != nil {
+			// not an official release if it's not semver compatiable.
+			continue
+		}
+		vs = append(vs, v)
+	}
+
+	// decending order from most recent.
+	sort.Sort(sort.Reverse(semver.Collection(vs)))
+
+	for _, v := range vs {
+		if v.LessThan(tagSemver) {
+			r.previousTag = fmt.Sprintf("v%v", v)
+			break
+		}
+	}
+
+	if r.previousTag == "" {
+		log.Printf("No previous release tag found for tag [%s]", r.tag)
+	} else {
+		log.Printf("Previous Tag [%s]", r.previousTag)
+	}
+
+	branches, err := r.getBranches()
+	if err != nil {
+		return err
+	}
+
+	var releaseBranch *github.Branch
+	for _, branch := range branches {
+		if branch.Name != nil && *branch.Name == expectedBranch {
+			releaseBranch = branch
+			break
+		}
+	}
+
+	if releaseBranch == nil {
+		return fmt.Errorf("release branch [%s] not found for new release [%s]", expectedBranch, r.tag)
+	}
+
+	r.tagBranch = expectedBranch
+	return nil
+}
+
+func (r *releaseData) cutNewTag() error {
+
+	// checkout remote branch
+	err := r.checkoutUpstream()
+	if err != nil {
+		return err
+	}
+
+	r.makeTag(r.tagBranch)
+
+	return nil
+}
+
+func (r *releaseData) printData() {
+
+	if r.dryRun {
+		log.Print("DRY-RUN")
+	}
+
+	log.Print("Input Data")
+
+	log.Printf("\trepoUrl: %s", r.repoUrl)
+	log.Printf("\tnewTag: %s", r.tag)
+	log.Printf("\tnewBranch: %s", r.newBranch)
+	log.Printf("\torg: %s", r.org)
+	log.Printf("\trepo: %s", r.repo)
+	log.Printf("\tforce: %t", r.force)
+	log.Printf("\tgithubTokenFile: %s", r.githubTokenPath)
+}
+
+func main() {
+	newBranch := flag.String("new-branch", "", "New branch to cut from master.")
+	releaseTag := flag.String("new-release", "", "New release tag. Must be a valid semver. The branch is automatically detected from the major and minor release")
+	org := flag.String("org", "", "The project org")
+	repo := flag.String("repo", "", "The project repo")
+	dryRun := flag.Bool("dry-run", true, "Should this be a dry run")
```

Comment:
I really like this option! I used this tool today to do the release and it gave me confidence to play around with the tool.

Reply:
yeah, i decided to leave dry-run on by default just in case, so it forces people to disable dry-run to do anything